src/java/org/apache/cassandra/db/compaction/unified/RealEnvironment.java

```diff
@@ -71,7 +71,7 @@ public double cacheMissRatio()
     @Override
     public double bloomFilterFpRatio()
     {
-        return metrics().bloomFilterFalseRatio.getValue();
+        return metrics() == null ? 0.0 : metrics().bloomFilterFalseRatio.getValue();
```

Comment:
This is weird... The docs of `CompactionRealm` say `metrics()` can be `null`, but `CFS` uses a non-null `final` field for metrics. So why do we have `null` here?

Reply:
This only happens during construction of the CFS. There is a really shit race condition in CFS that actually has a test for it. The issue is that lots of elements that are created in the CFS constructor and a number of these rely on the CFS metrics object. The problem is that the `TableMetrics` rely on a number of elements that provided in the constructor. I don't see any alternative way to fix this. We would end up rewriting the CFS construction, which I'd prefer not to do.